File: python/13.py
import re
import sys
from pathlib import Path
from typing import Any, List, Optional, Tuple

sys.path.append(str(Path(__file__).parent.parent.parent))
from Modules.timer import Timer
def part1(data):
    """ 2021 Day 13 Part 1

    >>> part1(['6,10', '0,14', '9,10', '0,3', '10,4', '4,11', '6,0', '6,12', '4,1', '0,13', '10,12', '3,4', '3,0', '8,4', '1,10', '2,14', '8,10', '9,0', '', 'fold along y=7', 'fold along x=5'])
    17
    """

    points = []
    for line in data:
        if len(line) == 0:
            continue

        if line[0] == 'f':
            info = line.split("fold along ")
            info = info[1:]
            info = info[0].split('=')
            info[-1] = int(info[-1])

            if info[0] == 'x':
                newPoints = []
                for point in points:
                    newPoint = [info[1] - abs(point[0] - info[1]), point[1]]
                    if not newPoint in newPoints:
                        newPoints.append(newPoint)

                points = newPoints[:]
            else:
                newPoints = []
                for point in points:
                    newPoint = [point[0], info[1] - abs(point[1] - info[1])]
                    if not newPoint in newPoints:
                        newPoints.append(newPoint)

                points = newPoints[:]

            break 
        else:
            digits = [int(num) for num in line.split(',')]
            if not digits in points:
                points.append(digits)

    return len(points)  


def part2(data):
    """ 2021 Day 13 Part 2
    """

    points = []
    for line in data:
        if len(line) == 0:
            continue

        if line[0] == 'f':
            info = line.split("fold along ")
            info = info[1:]
            info = info[0].split('=')
            info[-1] = int(info[-1])

            if info[0] == 'x':
                newPoints = []
                for point in points:
                    newPoint = [info[1] - abs(point[0] - info[1]), point[1]]
                    if not newPoint in newPoints:
                        newPoints.append(newPoint)

                points = newPoints[:]
            else:
                newPoints = []
                for point in points:
                    newPoint = [point[0], info[1] - abs(point[1] - info[1])]
                    if not newPoint in newPoints:
                        newPoints.append(newPoint)

                points = newPoints[:]       
        else:
            digits = [int(num) for num in line.split(',')]
            if not digits in points:
                points.append(digits)

    mins = [float('inf')] * 2
    maxs = [float('-inf')] * 2
    for point in points:
        for (i, j) in enumerate(point):
            if j > maxs[i]:
                maxs[i] = j
            
            if j < mins[i]:
                mins[i] = j
    
    string = ''
    for y in range(mins[1], maxs[1] + 1):
        string += '\n'
        for x in range(mins[0], maxs[0] + 1):
            string += '█' if [x, y] in points else ' '

    return string


def main(input_path: Optional[Path | str]=None, verbose: bool=False) -> Tuple[Tuple[Any, float]]:
    if not input_path:
        if not (input_path := sys.argv[1] if len(sys.argv) > 1 else None):
            year, day = re.findall(r'\d+', str(__file__))[-2:]
            input_path = Path(Path(__file__).parent.parent.parent, "Inputs", f"{year}_{day}.txt")
    
    with open(input_path, encoding='UTF-8') as f:
        data = [line.strip('\n') for line in f.readlines()]

    with Timer() as p1_time:
        p1 = part1(data)

    if verbose:
        print(f"\nPart 1:\nNumber of dots visible after first fold: {p1}\nRan in {p1_time.elapsed:0.4f} seconds")

    with Timer() as p2_time:
        p2 = part2(data)

    if verbose:
        print(f"\nPart 2:\nCode: {p2}\nRan in {p2_time.elapsed:0.4f} seconds")

    return [(p1, p1_time.elapsed), (p2, p2_time.elapsed)]


if __name__ == "__main__":
    main(verbose=True)

File: c/13.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <stdbool.h>
#include <ctype.h>
#include <math.h>
#define defaultInput "../../Inputs/2021_13.txt"
#include "../../Modules/input.h"
#include "../../Modules/set.h"


char *pointStr(int x, int y) {
    int sizex = ceil(log10(abs(x) + 1)), sizey = ceil(log10(abs(y) + 1));
    char *str = (char*)calloc(sizex + sizey + 2, sizeof(char));
    sprintf(str, "%d,%d", x, y);
    return str;
}

int part1(char *fileName) {
    struct vector *input_data = multiLine(fileName);

    struct set *points = createSet(stringsize, copyElement);
    int i = 0;
    for (i = 0; i < input_data->len; i++) {
        char *line = (char*)input_data->arr[i];
        if (strlen(line) == 0) {
            break;
        }

        addSet(points, line);
    }

    i++;

    int minx, maxx, miny, maxy;
    struct set *newPoints = createSet(stringsize, copyElement);
    char *instruction = (char*)copyElement(input_data->arr[i], stringsize(input_data->arr[i])), *p = strtok(instruction, "=");
    char dir = p[strlen(p) - 1];
    p = strtok(NULL, "=");
    int fold = atoi(p);

    struct vector *folding = set2vector(points);
    for (int k = 0; k < folding->len; k++) {
        char *point = (char*)copyElement(folding->arr[k], stringsize(folding->arr[k]));
        int x, y;

        p = strtok(point, ",");
        x = atoi(p);
        p = strtok(NULL, ",");
        y = atoi(p);

        if (dir == 'x') {
            if (x > fold) {
                x = fold - (x - fold);
            }
        } else {
            if (y > fold) {
                y = fold - (y - fold);
            }
        }

        if (k == 0) {
            minx = x;
            maxx = x;
            miny = y;
            maxy = y;
        } else {
            if (x < minx) {
                minx = x;
            }

            if (x > maxx) {
                maxx = x;
            }

            if (y < miny) {
                miny = y;
            }

            if (y > maxy) {
                maxy = y;
            }
        }

        addSet(newPoints, pointStr(x, y));
    }

    deleteSet(points, true);
    points = newPoints;

    return points->len;
}


char *part2(char *fileName) {
    struct vector *input_data = multiLine(fileName);

    struct set *points = createSet(stringsize, copyElement);
    int i = 0;
    for (i = 0; i < input_data->len; i++) {
        char *line = (char*)input_data->arr[i];
        if (strlen(line) == 0) {
            break;
        }

        addSet(points, line);
    }

    i++;

    int minx, maxx, miny, maxy;
    for (int j = 0; j + i < input_data->len; j++) {
        struct set *newPoints = createSet(stringsize, copyElement);
        char *instruction = (char*)copyElement(input_data->arr[i + j], stringsize(input_data->arr[i + j])), *p = strtok(instruction, "=");
        char dir = p[strlen(p) - 1];
        p = strtok(NULL, "=");
        int fold = atoi(p);

        struct vector *folding = set2vector(points);
        for (int k = 0; k < folding->len; k++) {
            char *point = (char*)copyElement(folding->arr[k], stringsize(folding->arr[k]));
            int x, y;

            p = strtok(point, ",");
            x = atoi(p);
            p = strtok(NULL, ",");
            y = atoi(p);

            if (dir == 'x') {
                if (x > fold) {
                    x = fold - (x - fold);
                }
            } else {
                if (y > fold) {
                    y = fold - (y - fold);
                }
            }

            if (k == 0) {
                minx = x;
                maxx = x;
                miny = y;
                maxy = y;
            } else {
                if (x < minx) {
                    minx = x;
                }

                if (x > maxx) {
                    maxx = x;
                }

                if (y < miny) {
                    miny = y;
                }

                if (y > maxy) {
                    maxy = y;
                }
            }

            addSet(newPoints, pointStr(x, y));
        }

        deleteSet(points, true);
        points = newPoints;
    }

    char *res = (char*)calloc((maxy - miny + 2) * (maxx - minx + 2), sizeof(char));
    for (int y = miny; y <= maxy; y++) {
        strcat(res, "\n");

        for (int x = minx; x <= maxx; x++) {
            strcat(res, (inSet(points, pointStr(x, y))) ? "█" : " ");
        }
    }

    return res;
}


int main (int argc, char *argv[]) {
    char *inputPath = defaultInput;
    if (argc > 1) {
        inputPath = argv[1];
    }

    clock_t t;
    t = clock(); 
    int p1 = part1(inputPath);
    t = clock() - t; 
    double t_p1 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 1:\nNumber of dots after first fold: %d\nRan in %f seconds\n", p1, t_p1);

    t = clock(); 
    char *p2 = part2(inputPath);
    t = clock() - t;
    double t_p2 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 2:\nMessage: %s\nRan in %f seconds\n", p2, t_p2);

    return 0;
}